sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,514 @@
+	if err != nil {
+		return 0, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetSupportedAttKeyIDNumRes = &pb.AesmServiceResponse_GetSupportedAttKeyIDNumResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return 0, err
+	}
+
+	if resp.GetSupportedAttKeyIDNumRes.GetErrorCode() != 0 {
+		return 0, fmt.Errorf("failed to get Supported Att Key ID Num (error code = %d)",
+			resp.GetSupportedAttKeyIDNumRes.GetErrorCode())
+	}
+
+	if resp.GetSupportedAttKeyIDNumRes.GetAttKeyIdNum() == 0 {
+		return 0, fmt.Errorf("invalid Att key Id Num = %d",
+			resp.GetSupportedAttKeyIDNumRes.GetAttKeyIdNum())
+	}
+
+	return resp.GetSupportedAttKeyIDNumRes.GetAttKeyIdNum(), nil
+}
+
+func getSupportedAttKeyIDsRequest(num uint32) ([]byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.GetSupportedAttKeyIDsReq = &pb.AesmServiceRequest_GetSupportedAttKeyIDsRequest{
+		BufSize: num * attestationKeyIdentityLength,
+		Timeout: timeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetSupportedAttKeyIDsRes = &pb.AesmServiceResponse_GetSupportedAttKeyIDsResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return nil, err
+	}
+
+	if resp.GetSupportedAttKeyIDsRes.GetErrorCode() != 0 {
+		return nil, fmt.Errorf("failed to get Supported Att Key IDs List (error code = %d)",
+			resp.GetSupportedAttKeyIDsRes.GetErrorCode())
+	}
+
+	if resp.GetSupportedAttKeyIDsRes.GetAttKeyIds() == nil {
+		return nil, fmt.Errorf("failed to get Supported Att Key IDs List")
+	}
+
+	return resp.GetSupportedAttKeyIDsRes.GetAttKeyIds(), nil
+}
+
+func getAttKeyId(quoteType string) ([]byte, error) {
+	if !strings.EqualFold(quoteType, QuoteTypeEcdsa) && !strings.EqualFold(quoteType, QuoteTypeEpidUnlinkable) && !strings.EqualFold(quoteType, QuoteTypeEpidLinkable) {
+		return nil, fmt.Errorf("Unsupport quote type: %v", quoteType)
+	}
+
+	attKeyId, err := selectAttKeyIDRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	akId := &attestationKeyIdentity{}
+	if err = restruct.Unpack(attKeyId, binary.LittleEndian, &akId); err != nil {
+		return nil, err
+	}
+
+	if strings.EqualFold(quoteType, QuoteTypeEcdsa) && (akId.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP256 || akId.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP384) {
+		return attKeyId, nil
+	} else if (strings.EqualFold(quoteType, QuoteTypeEpidUnlinkable) || strings.EqualFold(quoteType, QuoteTypeEpidLinkable)) && akId.AlgorithmId == sgxQuoteLibraryAlgorithmEpid {
+		return attKeyId, nil
+	}
+
+	akIdNum, err := getSupportedAttKeyIDNumRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	akIdList, err := getSupportedAttKeyIDsRequest(akIdNum)
+	if err != nil {
+		return nil, err
+	}
+
+	var i uint32 = 0
+	for i < akIdNum {
+		attKeyId = akIdList[i*attestationKeyIdentityLength : (i+1)*attestationKeyIdentityLength-1]
```

Comment:
Should be:
```
attKeyId = akIdList[i*attestationKeyIdentityLength : (i+1)*attestationKeyIdentityLength]
```

Other the length of attKeyId will be not attestationKeyIdentityLength.

Reply:
> Should be:
> 
> ```
> attKeyId = akIdList[i*attestationKeyIdentityLength : (i+1)*attestationKeyIdentityLength]
> ```
> 
> Other the length of attKeyId will be not attestationKeyIdentityLength.

OK, thanks, it's a bug.